unittests/Interpreter/InterpreterInfra.h

```diff
@@ -0,0 +1,153 @@
+// Copyright 2019 The Souper Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOUPER_INTERPRTER_INFRA_H
+#define SOUPER_INTERPRTER_INFRA_H
+
+#include "llvm/Support/KnownBits.h"
+#include "llvm/IR/ConstantRange.h"
+
+#include "souper/Infer/Interpreter.h"
+#include "souper/Infer/AbstractInterpreter.h"
+
+#include <bitset>
+
+// width used for transfer function tests
+constexpr int WIDTH = 4;
+
+namespace souper {
+
+    struct EvalValueKB : public EvalValue {
+      llvm::KnownBits ValueKB;
+
+      EvalValueKB(llvm::KnownBits Val);
+      EvalValueKB(EvalValue &Val);
+      llvm::KnownBits getValueKB();
+    };
+
+    class KBTesting {
+      // wrapper over @llvm::KnownBitsAnalysis::mergeKnownBits()
+      EvalValueKB merge(EvalValueKB a, EvalValueKB b);
+
+      llvm::KnownBits setLowest(llvm::KnownBits x);
+      llvm::KnownBits clearLowest(llvm::KnownBits x);
+
+      EvalValueKB bruteForce(llvm::KnownBits x, llvm::KnownBits y, Inst::Kind Pred);
+    public:
+      static bool nextKB(llvm::KnownBits &x);
+      bool testFn(Inst::Kind pred);
+    };
+
+
+
+    class CRTesting {
+      bool ok(const llvm::ConstantRange &R, const bool Table[]);
+
+      // Find the largest hole and build a llvm::ConstantRange around it
+      llvm::ConstantRange bestCR(const bool Table[], const int Width);
+      llvm::ConstantRange exhaustive(const llvm::ConstantRange &L, const llvm::ConstantRange &R,
+			       Inst::Kind pred, const llvm::ConstantRange &Untrusted);
+
+      void check(const llvm::ConstantRange &L, const llvm::ConstantRange &R, Inst::Kind pred,
+		 double &FastBits, double &PreciseBits, int &Count, int &PreciseCount);
+
+    public:
+
+      static llvm::ConstantRange nextCR(const llvm::ConstantRange &CR);
+      bool testFn(Inst::Kind pred);
+    };
+
+ namespace TestingUtil {
+
+    void exhaustiveKBCRReduction(llvm::KnownBits &KB, llvm::ConstantRange &CR);
+
+    template<int SetSize>
+    std::bitset<SetSize> concretizeKB(const llvm::KnownBits &KB) {
+      assert(SetSize == (1 << KB.getBitWidth()));
+
+      std::bitset<SetSize> ResultSet(1);
+      for (int Cnt = 0; Cnt < SetSize; Cnt++) {
+
+	llvm::KnownBits ValueKB;
+	auto Tmp = llvm::APInt(KB.getBitWidth(), Cnt);
+	ValueKB.One = Tmp;
+	ValueKB.Zero = ~Tmp;
+
+	if (KnownBitsAnalysis::isConflictingKB(ValueKB, KB))
```

Comment:
how can there be a conflict given that you initialized One to Tmp and Zero to ~Tmp?

Reply:
`Tmp` and `ValueKB` is the APInt, KnownBits representation respectively of the number `Cnt`. I am checking if this KnownBits representation of the number conflicts with the `KB` which is the KnownBits object we want to concretize. 

There would be some value of `Cnt` which will not agree with the `KnownBits` value that we are concretizing.